neqo-transport/src/stream_id.rs

```diff
@@ -127,52 +105,172 @@ impl ::std::fmt::Display for StreamId {
     }
 }
 
-#[derive(Debug, Eq, PartialEq, Clone, Copy, Ord, PartialOrd, Hash)]
-pub struct StreamIndex(u64);
-
-impl StreamIndex {
-    pub const fn new(val: u64) -> Self {
-        Self(val)
-    }
+pub struct RemoteStreamsFlowControl {
+    streams_fc: ReceiverFlowControl<StreamType>,
+    next_stream: u64,
+    stream_type_bits: u64,
+}
 
-    pub fn to_stream_id(self, stream_type: StreamType, role: Role) -> StreamId {
+impl RemoteStreamsFlowControl {
+    pub fn new(stream_type: StreamType, max_streams: u64, role: Role) -> Self {
         let type_val = match stream_type {
             StreamType::BiDi => 0,
             StreamType::UniDi => 2,
         };
+        // This is for a stream created by a peer, therefore the role bit is opposit.
         let role_val = match role {
-            Role::Server => 1,
-            Role::Client => 0,
+            Role::Server => 0,
+            Role::Client => 1,
         };
 
-        StreamId::from((self.0 << 2) + type_val + role_val)
+        Self {
+            streams_fc: ReceiverFlowControl::new(stream_type, max_streams),
+            next_stream: 0,
+            stream_type_bits: type_val + role_val,
+        }
     }
 
-    pub fn as_u64(self) -> u64 {
-        self.0
+    pub fn write_frames(
+        &mut self,
+        builder: &mut PacketBuilder,
+        tokens: &mut Vec<RecoveryToken>,
+        stats: &mut FrameStats,
+    ) -> Res<()> {
+        self.streams_fc.write_frames(builder, tokens, stats)
+    }
+
+    pub fn is_new_stream(&mut self, stream_id: StreamId) -> Res<bool> {
+        let stream_idx = stream_id.as_u64() >> 2;
+        if !self.streams_fc.check_allowed(stream_idx) {
+            return Err(Error::StreamLimitError);
+        }
+        Ok(stream_idx >= self.next_stream)
+    }
+
+    pub fn add_stream(&mut self) -> Option<StreamId> {
+        let new_stream = self.next_stream;
+        self.next_stream += 1;
+        assert!(self.streams_fc.check_allowed(new_stream));
+        Some(StreamId::from((new_stream << 2) + self.stream_type_bits))
+    }
+
+    pub fn send_flowc_update(&mut self) {
+        self.streams_fc.send_flowc_update();
+    }
+
+    pub fn max_streams_lost(&mut self, max_streams: u64) {
+        self.streams_fc.lost(max_streams);
+    }
+
+    pub fn add_retired(&mut self, removed: u64) {
+        self.streams_fc.add_retired(removed);
+        if removed > 0 {
+            // Send a update immediately.
+            self.streams_fc.send_flowc_update();
+        }
+    }
+}
+
+pub struct RemoteStreamsFlowControls([RemoteStreamsFlowControl; 2]);
```

Comment:
I think that it might be better to have a StreamLimits object that has both local and remote limits all together.  That should take more of the load from Connection, which has a lot of work to do.

I was considering whether RecvStreams and SendStreams could handle this, but you are right to not put anything there; these limits don't work there as they deal with stream components, not the entire stream.

A wrapper means that you can't use Index[Mut], but I think that is OK.  `c.stream_limits.local_mut(UniDi)` and `c.stream_limits.remote(BiDi)` are pretty clear still.

Also, you don't need to use a fixed-size array for this.  It's OK to use named variables.

Reply:
First I implemented StreamLimits (I think it really had the same name), but I did not like it. Somehow it felt like a half a way and had too complex interface. I think I wanted to move stream creation into that function. 
I need to think about it, maybe we need Streams object that holds all streams, the limits and it is responsible for creating streams and remove them. To make the usage easier recv_streams and send_streams can be public.